js-raphtory/src/graph/mod.rs

```diff
@@ -0,0 +1,250 @@
+use std::ops::Deref;
+use std::sync::Arc;
+
+use js_sys::Array;
+use js_sys::Object;
+use raphtory::core::tgraph_shard::errors::GraphError;
+use raphtory::core::Prop;
+use raphtory::db::graph::Graph as TGraph;
+use raphtory::db::graph_window::WindowedGraph;
+use raphtory::db::view_api::internal::GraphViewInternalOps;
+use raphtory::db::view_api::GraphViewOps;
+use raphtory::db::view_api::TimeOps;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+
+use chrono::{Datelike, Timelike};
+
+use crate::graph::vertex::JsVertex;
+use crate::graph::vertex::Vertex;
+use crate::log;
+use crate::utils::set_panic_hook;
+
+mod edge;
+mod graph_view_impl;
+mod vertex;
+
+#[derive(Clone, Debug)]
+enum UnderGraph {
+    TGraph(Arc<TGraph>),
+    WindowedGraph(Arc<WindowedGraph<TGraph>>),
+}
+
+impl UnderGraph {
+    pub fn mutable_graph(&self) -> &TGraph {
+        match self {
+            UnderGraph::TGraph(g) => g,
+            UnderGraph::WindowedGraph(g) => &g.graph,
+        }
+    }
+
+    // a bit heavy but might work
+    pub fn graph(&self) -> Box<Arc<dyn GraphViewInternalOps + Send + Sync + 'static>> {
+        match self {
+            UnderGraph::TGraph(g) => Box::new(g.clone()),
+            UnderGraph::WindowedGraph(g) => Box::new(g.clone()),
+        }
+    }
+}
+
+#[wasm_bindgen]
+#[repr(transparent)]
+#[derive(Clone)]
+pub struct Graph(UnderGraph);
+
+#[wasm_bindgen]
+#[derive(Debug)]
+pub struct JSError(GraphError);
+
+struct JsObjectEntry(JsValue);
+
+#[repr(transparent)]
+pub(crate) struct JsProp(Prop);
+
+impl Into<JsValue> for JsProp {
+    fn into(self) -> JsValue {
+        match self.0 {
+            raphtory::core::Prop::Str(v) => v.into(),
+            raphtory::core::Prop::I32(v) => v.into(),
+            raphtory::core::Prop::I64(v) => v.into(),
+            raphtory::core::Prop::U32(v) => v.into(),
+            raphtory::core::Prop::U64(v) => v.into(),
+            raphtory::core::Prop::F32(v) => v.into(),
+            raphtory::core::Prop::F64(v) => v.into(),
+            raphtory::core::Prop::Bool(v) => v.into(),
+            raphtory::core::Prop::DTime(v) => {
+                js_sys::Date::new_with_year_month_day_hr_min_sec_milli(
+                    v.year() as u32,
+                    v.month() as i32,
+                    v.day() as i32,
+                    v.hour() as i32,
+                    v.minute() as i32,
+                    v.second() as i32,
+                    0,
+                )
+                .into()
+            }
+        }
+    }
+}
+
+impl Deref for JsProp {
+    type Target = Prop;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<JsObjectEntry> for Option<(String, Prop)> {
+    fn from(entry: JsObjectEntry) -> Self {
+        let arr: Array = entry.0.into();
+
+        let key = arr.at(0).as_string().unwrap();
+        let value = arr.at(1).as_string().unwrap();
+        Some((key, Prop::Str(value)))
+    }
+}
+
+#[wasm_bindgen]
+impl Graph {
```

Comment:
Probably doesn't matter for now, but we should probably break these out into their own files

Reply:
I moved everything outside, Graph is the only one in mod.rs

Reply:
Any reason for keeping it in there? Could just move it out.

Reply:
yes it's easy in js to just do import it
```import { Graph } from "js-raphtory";````